gtsfm/bundle/rig_bundle_adjustment.py

```diff
@@ -0,0 +1,186 @@
+
+METRICS_PATH: Path = Path(__file__).resolve().parent.parent.parent / "result_metrics"
+
+TEST_DATA_ROOT: Path = Path(__file__).resolve().parent.parent.parent / "tests" / "data"
+HILTI_TEST_DATA_PATH: Path = TEST_DATA_ROOT / "hilti_exp4_small"
+
+"""In this file, we use the GTSAM's GeneralSFMFactor2 instead of GeneralSFMFactor because Factor2 enables decoupling
+of the camera pose and the camera intrinsics, and hence gives an option to share the intrinsics between cameras.
+"""
+X = symbol_shorthand.X  # camera pose
+B = symbol_shorthand.B  # body frame (IMU) pose
+
+CAM_POSE3_DOF = 6  # 6 dof for pose of camera
+
+# noise model params
+CAM_POSE3_PRIOR_NOISE_SIGMA = 0.1
+CAM_IMU_POSE_PRIOR_SIGMA = 1e-3
+
+logger = logger_utils.get_logger()
+
+
+class RigBundleAdjustmentOptimizer(BundleAdjustmentOptimizer):
+    """Bundle adjustment for special rig multi-sensory devices."""
+
+    # TODO(Ayush): share calibrations between same cams?
+    # TODO(Ayush): use a diagonal model for calibration prior as distortion should have much lower sigmas?
+
+    def __init__(
+        self,
+        output_reproj_error_thresh: Optional[float] = None,
+        robust_measurement_noise: bool = False,
+        shared_calib: bool = False,
+        max_iterations: Optional[int] = None,
+    ) -> None:
+        super().__init__(output_reproj_error_thresh, robust_measurement_noise, shared_calib, max_iterations)
+
+        # Temporary hack to get relative pose between cameras and IMUs
+        hilti_loader = HiltiLoader(
+            base_folder=str(HILTI_TEST_DATA_PATH),
+            max_frame_lookahead=1,
+            max_length=None,
+        )
+
+        self._cam_T_imu = hilti_loader.get_camTimu()
+
+    def __get_rig_idx(self, camera_idx: int) -> int:
+        return camera_idx // 5
+
+    def __get_camera_type(self, camera_idx: int) -> int:
+        return camera_idx % 5
+
+    def _between_factors(
+        self, relative_pose_priors: Dict[Tuple[int, int], Optional[PosePrior]], cameras_to_model: List[int]
+    ) -> NonlinearFactorGraph:
+        """Add between factors on poses between cameras and IMUs.
+
+        1. For the same timestamp, each camera has a BetweenFactor to the IMU at the same timestamp.
+        2. For different timestamps, the between factors are between IMUs.
+        """
+        graph = NonlinearFactorGraph()
+
+        # add a between factor for each camera with its IMU
+        for i in cameras_to_model:
+            camera_type: int = self.__get_camera_type(i)
+            rig_idx: int = self.__get_rig_idx(i)
+            graph.push_back(
+                BetweenFactorPose3(
+                    X(i),
+                    B(rig_idx),
+                    self._cam_T_imu[camera_type],
+                    gtsam.noiseModel.Isotropic.Sigma(CAM_POSE3_DOF, CAM_IMU_POSE_PRIOR_SIGMA),
+                )
+            )
+
+        # translate the relative pose priors between cams to IMUs, and add if not already present
```

Comment:
Why do we translate them? Camera poses are also variables, right?

Reply:
I get it now. But we should get them from the constraints file.

Reply:
Resolved as we do not use IMU anymore.